contracts/token/ERC721/ERC721.sol

```diff
@@ -340,8 +350,15 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata {
         // Clear approvals from the previous owner
         _approve(address(0), tokenId);
 
-        _balances[from] -= 1;
-        _balances[to] += 1;
+        unchecked {
+            // `_balances[from]` cannot overflow for the same reason as described in `_burn`:
+            // `from`'s balance is the number of token held, which is at least one before the current
+            // transfer.
+            // `_balances[to]` could overflow in the conditions described in `_mint`. That would require
+            // all 2**256 token ids to be minted, which in practice is impossible.
+            _balances[from] -= 1;
```

Comment:
`_beforeTokenTransfer` could in theory move the token away and this would result in overflow if we use unchecked.

```suggestion
        _balances[from] -= 1;
        unchecked {
            // `_balances[to]` could overflow in the conditions described in `_mint`. That would require
            // all 2**256 token ids to be minted, which in practice is impossible.
```

Reply:
If `_beforeTokenTransfer` transfers the token, then the balances would be invalid regardless of the `unchecked`. This is a wider issue, that we should address independently, and that IMO should not prevent the use of unchecked here.